platform/linux-generic/odp_packet_io.c

```diff
@@ -89,6 +96,274 @@ int odp_pktio_init_local(void)
 	return odp_pktio_ops_init_local(true);
 }
 
+static int write_pcapng_hdr(pktio_entry_t *entry, int qidx)
+{
+	size_t len;
+	pcapng_section_hdr_block_t shb;
+	pcapng_interface_description_block_s idb;
+	int fd = entry->s.pcapng_info.pcapng_fd[qidx];
+
+	memset(&shb, 0, sizeof(shb));
+	memset(&idb, 0, sizeof(idb));
+
+	shb.block_type = PCAPNG_BLOCK_TYPE_SHB;
+	shb.block_total_length = sizeof(shb);
+	shb.block_total_length2 = sizeof(shb);
+	shb.magic = PCAPNG_ENDIAN_MAGIC;
+	shb.version_major = 0x1;
+	shb.version_minor = 0x0;
+	shb.section_len = -1;
+
+	len = write(fd, &shb, sizeof(shb));
+	fsync(fd);
+	/* fail to write shb/idb means the pcapng is unreadable */
+	if (!len) {
+		ODP_ERR("Failed to write pcapng section hdr\n");
+		close(fd);
+		return -1;
+	}
+
+	idb.block_type = PCAPNG_BLOCK_TYPE_IDB;
+	idb.block_total_length = sizeof(idb);
+	idb.block_total_length2 = sizeof(idb);
+	idb.linktype = 0x1; /* LINKTYPE_ETHERNET */
+	idb.snaplen = 0x0; /* unlimited */
+	len = write(fd, &idb, sizeof(idb));
+	if (!len) {
+		ODP_ERR("Failed to write pcapng interface description\n");
+		close(fd);
+		return -1;
+	}
+	fsync(fd);
+
+	entry->s.pcapng_info.state[qidx] = PCAPNG_WR_HDR;
+
+	return 0;
+}
+
+static int write_pcapng_pkts(pktio_entry_t *entry, int qidx,
+			     const odp_packet_t packets[], int num)
+{
+	int i = 0;
+	struct iovec packet_iov[3 * num];
+	pcapng_enhanced_packet_block_t epb;
+	int iovcnt = 0;
+	ssize_t len = 0;
+	int fd = entry->s.pcapng_info.pcapng_fd[qidx];
+
+	for (i = 0; i < num; i++) {
+		odp_packet_hdr_t *pkt_hdr = odp_packet_hdr(packets[i]);
+		uint32_t pkt_len = _odp_packet_len(packets[i]);
+		char *buf = (char *)odp_packet_data(packets[i]);
+
+		epb.block_type = PCAPNG_BLOCK_TYPE_EPB;
+		epb.block_total_length = sizeof(epb) +
+			ROUNDUP_ALIGN(pkt_len, PCAP_DATA_ALIGN) +
+			PCAP_DATA_ALIGN;
+		epb.interface_idx = 0;
+		epb.timestamp_high = (uint32_t)(pkt_hdr->timestamp.u64 >> 32);
+		epb.timestamp_low = (uint32_t)(pkt_hdr->timestamp.u64);
+		epb.captured_len = pkt_len;
+		epb.packet_len = pkt_len;
+
+		/* epb */
+		packet_iov[iovcnt].iov_base = &epb;
+		packet_iov[iovcnt].iov_len = sizeof(epb);
+		iovcnt++;
+
+		/* data */
+		packet_iov[iovcnt].iov_base = buf;
+		packet_iov[iovcnt].iov_len =
+			ROUNDUP_ALIGN(pkt_len, PCAP_DATA_ALIGN);
+		iovcnt++;
+
+		/* trailing */
+		packet_iov[iovcnt].iov_base = &epb.block_total_length;
+		packet_iov[iovcnt].iov_len = sizeof(uint32_t);
+		iovcnt++;
+	}
+
+	/* we don't really care if we manage to write *all* data */
```

Comment:
In fact we do care. We can't tolerate partial packet writes, otherwise pcapng file will be unreadable. The proper way to address that is to see how much space is left in the pipe (buf_size - TIOCOUTQ) and append packets to iovec as long as they fit in the calculated 'budget'.

Another thing to consider -- is increasing pipe buffer size. The linux default size is 64K, so 'dd' command you used doesn't make much sense. Please consider setting it with fcntl to max size (1M according to /proc/sys/fs/pipe-max-size).